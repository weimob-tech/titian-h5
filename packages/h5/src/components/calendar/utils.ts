/* eslint-disable @titian/no-more-args */
import { getDate } from '../common/utils/index';
import type { CalDateProps, CalMonthProps, CalRenderDateProps, CalWeekProps } from './const';
import { CalTypeEnum, DateStatusEnum, WEEK } from './const';

export const DAYTIME = 1000 * 60 * 60 * 24; // 每天的毫秒数

export function getCalDate(date: Date) {
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getDate();
  const week = date.getDay();
  const formatStr = `${year}${month > 9 ? month : `0${month}`}${day > 9 ? day : `0${day}`}`;
  const format = Number(formatStr);
  const calendarDay: CalDateProps = {
    date,
    time: date.getTime(),
    year,
    month,
    day,
    week,
    fullDateNum: format,
    text: day,
    status: DateStatusEnum.NOT_STATUS,
    topInfo: '',
    bottomInfo: '',
  };
  return calendarDay;
}
export const getWeek = (startWeek = WEEK[0].name, week: CalWeekProps[] = WEEK): CalWeekProps[] => {
  const begin = startWeek || WEEK[0].name;
  const number = week.findIndex(item => item.name === begin);
  if (number === 0) {
    return week;
  }
  const end = week.slice(0, number);
  const start = week.slice(number, week.length);
  return [...start, ...end];
};
export function diff(before: CalDateProps, after: CalDateProps) {
  return Math.abs((after.time - before.time) / DAYTIME) + 1;
}
export function calcMonthData(
  { min, max }: { min: CalDateProps; max: CalDateProps },
  currentDate: CalDateProps[],
): [CalMonthProps[], CalRenderDateProps[]] {
  let number = (max.year - min.year) * 12 + (max.month - min.month) + 1;
  number = number > 0 ? number : 1;
  let id = '';
  const renderDate: CalRenderDateProps[] = [];
  const monthList = new Array(number).fill(0).map((_item, index) => {
    const date = getDate(min.year, min.month - 1 + index);
    const [time, year, month] = [date.getTime(), date.getFullYear(), date.getMonth() + 1];
    currentDate.forEach(cur => {
      if (cur.year === year && cur.month === month) {
        if (!id) {
          id = `month${index}`;
        }
        renderDate.push({ ...cur, parentIndex: index });
      }
    });
    return { time, year, month };
  });
  return [monthList, renderDate];
}

/**
 * 是否需要覆盖
 * @param month  月份
 * @param select 选中值
 * @param mode 类型
 * @returns booolean
 */
export function judgeCover(month: CalRenderDateProps, select: CalRenderDateProps[], mode: `${CalTypeEnum}`): boolean {
  if (mode !== CalTypeEnum.RANGE) {
    return false;
  }
  if (select.length !== 2) {
    return false;
  }
  const [start, end] = select;
  const startIndex = start.parentIndex;
  const endIndex = end.parentIndex;
  const number = month.parentIndex;
  return number > startIndex && number < endIndex;
}

/**
 * 是否需要计算
 * @param month  月份
 * @param select 选中值
 * @param mode 类型
 * @returns booolean
 */
export function judgeCompare(month: CalRenderDateProps, select: CalRenderDateProps[], mode: `${CalTypeEnum}`): boolean {
  if (mode !== CalTypeEnum.RANGE) {
    return false;
  }
  return select.some(i => i.parentIndex === month.parentIndex);
}

function getCalculateStatus(
  day: CalDateProps,
  border: { min?: CalDateProps; max?: CalDateProps },
  select: CalDateProps[],
): DateStatusEnum {
  const selected = select.find(i => i.fullDateNum === day.fullDateNum);
  if (selected) {
    return selected.status;
  }
  let status = DateStatusEnum.NOT_STATUS;
  if (border.min && border.max) {
    status =
      border.min.fullDateNum > day.fullDateNum || day.fullDateNum > border.max.fullDateNum
        ? DateStatusEnum.DISABLED
        : DateStatusEnum.NOT_STATUS;
  } else {
    if (border.min) {
      status = border.min.fullDateNum > day.fullDateNum ? DateStatusEnum.DISABLED : DateStatusEnum.NOT_STATUS;
    }
    if (border.max) {
      status = day.fullDateNum > border.max.fullDateNum ? DateStatusEnum.DISABLED : DateStatusEnum.NOT_STATUS;
    }
  }
  return status;
}

const getRenderDateByCalendar = ({
  calDate,
  select,
  border,
  index,
  formatter,
}: {
  calDate: CalDateProps;
  select: CalDateProps[];
  border: { max?: CalDateProps; min?: CalDateProps };
  index: number;
  formatter?: (day: CalRenderDateProps) => CalRenderDateProps;
}): CalRenderDateProps => {
  const status = getCalculateStatus(calDate, border, select);
  const calRenderDate = {
    ...calDate,
    status,
    parentIndex: index,
  };
  if (typeof formatter === 'function') {
    if (calRenderDate.status === DateStatusEnum.RANGE_FULL) {
      const startDate = formatter({
        ...calRenderDate,
        status: DateStatusEnum.RANGE_START,
      });
      const endDate = formatter({
        ...calRenderDate,
        status: DateStatusEnum.RANGE_END,
      });
      const bottomInfo = [startDate.bottomInfo, endDate.bottomInfo].map(i => i).join('/');
      calRenderDate.bottomInfo = bottomInfo;
      return calRenderDate;
    }
    return formatter(calRenderDate);
  }
  return calRenderDate;
};

export function getDayByMonth({
  month,
  select,
  border,
  formatter,
  parentIndex,
}: {
  month: string | number | Date;
  select: CalDateProps[];
  border: { max?: CalDateProps; min?: CalDateProps };
  formatter?: (day: CalRenderDateProps) => CalRenderDateProps;
  parentIndex: number;
}): CalRenderDateProps[] {
  const date: Date = getDate(month);
  const end = getDate(date.getFullYear(), date.getMonth() + 1, 0);
  const number = end.getDate();
  return new Array(number).fill(null).map((_item, index) => {
    const calDate = getCalDate(getDate(date.getFullYear(), date.getMonth(), index + 1));
    return getRenderDateByCalendar({ calDate, select, border, index: parentIndex, formatter });
  });
}

export const mergeContinuousDate = <T extends CalDateProps>(list: T[]): T[] => {
  let temp: T[] = [list[0]];
  const result = [temp];
  let i = 0;
  if (list.length < 2) {
    return list.map(item => ({
      ...item,
      status: DateStatusEnum.MULTIPLE,
    }));
  }
  for (let j = 1; j < list.length; j += 1) {
    const last = getDate(list[i].year, list[i].month - 1, list[i].day + 1);
    const day = getDate(list[j].year, list[j].month - 1, list[j].day);
    const year = day.getFullYear();
    const month = day.getMonth();
    const date = day.getDate();
    const lastYear = last.getFullYear();
    const lastMonth = last.getMonth();
    const lastDate = last.getDate();
    if (year === lastYear && month === lastMonth && date === lastDate) {
      temp.push(list[j]);
    } else {
      temp = [list[j]];
      result.push(temp);
    }
    i = j;
  }
  const resetResult: T[] = [];

  for (let x = 0; x < result.length; x += 1) {
    const curList = result[x];
    const curLength = curList.length;
    curList.forEach((item, index) => {
      if (curLength === 1) {
        resetResult.push({ ...item, status: DateStatusEnum.MULTIPLE });
        return;
      }
      if (index === 0) {
        resetResult.push({ ...item, status: DateStatusEnum.MULTIPLE_START });
        return;
      }
      if (index === curLength - 1) {
        resetResult.push({ ...item, status: DateStatusEnum.MULTIPLE_END });
        return;
      }
      resetResult.push({ ...item, status: DateStatusEnum.MULTIPLE_MIDDLE });
    });
  }
  return resetResult;
};
