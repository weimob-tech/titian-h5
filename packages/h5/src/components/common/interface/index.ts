export type OmitType<T, P> = Pick<T, { [key in keyof T]: T[key] extends P ? never : key }[keyof T]>;

export type Key = string | number | symbol;

export type IAnyObject = { [key: Key]: unknown };

export type IAnyMap = Map<Key, unknown>;

export type DistanceUnit = 'px' | 'em' | 'rem' | 'vw' | 'vh' | '%';

export type IPosition = {
  bottom?: `${number}${DistanceUnit}`;
  top?: `${number}${DistanceUnit}`;
  left?: `${number}${DistanceUnit}`;
  right?: `${number}${DistanceUnit}`;
};
