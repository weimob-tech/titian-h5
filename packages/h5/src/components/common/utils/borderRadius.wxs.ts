function computedStyle(shape) {
  const style = {};
  if (shape === 'capsule' || shape === 'circle') {
    style['--base-radius-size'] = '999px';
  } else if (shape === 'rect' || shape === 'square') {
    style['--base-radius-size'] = '-999px';
  }
  return style;
}

export default {
  handler: computedStyle,
};
