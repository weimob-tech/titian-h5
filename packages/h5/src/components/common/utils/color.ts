/* eslint-disable no-nested-ternary */
/* eslint-disable prefer-regex-literals */
/* eslint-disable no-bitwise */
const HSBToRGB = (h: number, s: number, b: number): [number, number, number] => {
  s /= 100;
  b /= 100;
  const k = (n: number) => (n + h / 60) % 6;
  const f = (n: number) => b * (1 - s * Math.max(0, Math.min(k(n), 4 - k(n), 1)));
  const R = Math.round(255 * f(5));
  const G = Math.round(255 * f(3));
  const B = Math.round(255 * f(1));
  return [R, G, B];
};

const RGBToHSB = (r: number, g: number, b: number): [number, number, number] => {
  r /= 255;
  g /= 255;
  b /= 255;
  const v = Math.max(r, g, b);
  const n = v - Math.min(r, g, b);
  const h = n === 0 ? 0 : n && v === r ? (g - b) / n : v === g ? 2 + (b - r) / n : 4 + (r - g) / n;
  return [60 * (h < 0 ? h + 6 : h), v && (n / v) * 100, v * 100];
};

const hexToRGB = (hex: string, a?: number | string): string => {
  let alpha = false;
  let h: string | number = hex.slice(hex.startsWith('#') ? 1 : 0);
  if (h.length === 3) {
    h = [...h].map(x => x + x).join('');
  } else if (h.length === 8) {
    alpha = true;
  } else if (h.length === 6) {
    // TODO
  } else {
    return hex;
  }
  h = parseInt(h, 16);
  const isAlpha = alpha || a;
  a = a || `${h & 0x000000ff}`;
  return `rgb${isAlpha ? 'a' : ''}(${h >>> (alpha ? 24 : 16)}, ${
    (h & (alpha ? 0x00ff0000 : 0x00ff00)) >>> (alpha ? 16 : 8)
  }, ${(h & (alpha ? 0x0000ff00 : 0x0000ff)) >>> (alpha ? 8 : 0)}${isAlpha ? `, ${a}` : ''})`;
};

const RGBToHex = (r: number, g: number, b: number) => `#${((r << 16) + (g << 8) + b).toString(16).padStart(6, '0')}`;

const RGBAToHex = (rgb: string) => {
  const reg = new RegExp('^rgba?\\((.+)\\)$', 'g');

  if (!reg.test(rgb)) {
    return rgb;
  }

  const a = rgb.replace(reg, (_s, $1) => $1);
  const [r, g, b] = a.split(',').map(x => parseInt(x, 10));
  return RGBToHex(r, g, b);
};

const hexToHSB = (hex: string): [number, number, number] => {
  let alpha = false;
  let h: string | number = hex.slice(hex.startsWith('#') ? 1 : 0);
  if (h.length === 3) {
    h = [...h].map(x => x + x).join('');
  } else if (h.length === 8) {
    alpha = true;
  }
  h = parseInt(h, 16);
  const r = h >>> (alpha ? 24 : 16);
  const g = (h & (alpha ? 0x00ff0000 : 0x00ff00)) >>> (alpha ? 16 : 8);
  const b = (h & (alpha ? 0x0000ff00 : 0x0000ff)) >>> (alpha ? 8 : 0);
  return RGBToHSB(r, g, b);
};

const RGBToRGBA = (rgb: string, alpha = 1) => {
  const reg = new RegExp('^rgba?\\((.+)\\)$', 'g');

  if (!reg.test(rgb)) {
    return rgb;
  }

  const a = rgb.replace(reg, (_s, $1) => $1);
  const [r, g, b] = a.split(',').map(x => parseInt(x, 10));
  return `rgba(${r},${g},${b},${alpha})`;
};

const hexToRGBA = (hex: string, alpha = 1) => {
  const rgb = hexToRGB(hex);
  return RGBToRGBA(rgb, alpha);
};

const getNeutralsGrey = (precent = 85, diffs?: number[]) => {
  const group = ['#f2f2f2', '#f5f5f5', '#fafafa', '#ffffff'];
  let b = 0;
  const x = (3 * precent) / 17;

  function pipe(m: number, n: number) {
    b += m;
    const rgb = HSBToRGB(0, 0, b + n > 0 ? b + n : 0);
    return RGBToHex(...rgb);
  }

  const [d1 = 0, d2 = 0, d3 = 0, d4 = 0, d5 = 0] = diffs || [];
  const steps: [number, number][] = [
    [100 - precent, d1],
    [2 * x, d2],
    [x, d3],
    [x, d4],
    [x, d5],
  ];
  return steps.map(arr => pipe(...arr)).concat(group);
};
// console.log(getNeutralsGrey(85, [-2, 1, 2, 2, -2]));
// console.log(getNeutralsGrey(70));
// console.log(getNeutralsGrey(100));
export { HSBToRGB, RGBToHSB, hexToRGB, RGBToHex, hexToHSB, getNeutralsGrey, RGBAToHex, RGBToRGBA, hexToRGBA };
