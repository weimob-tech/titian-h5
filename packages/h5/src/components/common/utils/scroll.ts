import { raf } from './raf';
import { easeInOutCubic } from './index';

interface ScrollToOptions {
  getContainer?: () => HTMLElement | Window | Document;
  callback?: () => unknown;
  duration?: number;
}

function isWindow(obj: any) {
  return obj !== null && obj !== undefined && obj === obj.window;
}

export function getScroll(target: HTMLElement | Window | Document | null, top: boolean): number {
  const method = top ? 'scrollTop' : 'scrollLeft';
  let result = 0;
  if (isWindow(target)) {
    result = (target as Window)[top ? 'pageYOffset' : 'pageXOffset'];
  } else if (target instanceof Document) {
    result = target.documentElement[method];
  } else if (target) {
    result = (target as HTMLElement)[method];
  }

  if (target && !isWindow(target) && typeof result !== 'number') {
    result = ((target as HTMLElement).ownerDocument || (target as Document)).documentElement?.[method];
  }
  return result;
}

export function scrollTo(y: number, options: ScrollToOptions = {}) {
  const { getContainer = () => window, callback, duration = 450 } = options;

  const container = getContainer();
  const scrollTop = getScroll(container, true);
  const startTime = Date.now();

  const func = () => {
    const timestamp = Date.now();
    const time = timestamp - startTime;
    const nexScrollTop = easeInOutCubic(time > duration ? duration : time, scrollTop, y, duration);

    if (isWindow(container)) {
      (container as Window).scrollTo(window.pageXOffset, nexScrollTop);
    } else if (container instanceof Document || container.constructor.name === 'HTMLDocument') {
      (container as Document).documentElement.scrollTop = nexScrollTop;
    } else {
      (container as HTMLElement).scrollTop = nexScrollTop;
    }

    if (time < duration) {
      raf(func);
    } else if (typeof callback === 'function') {
      callback();
    }
  };

  raf(func);
}
