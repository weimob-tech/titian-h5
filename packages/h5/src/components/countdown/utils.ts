export interface TimeDataType {
  day: number;
  hour: number;
  minute: number;
  second: number;
  millisecond: number;
}
export function formatDuration(ms: number): TimeDataType {
  ms = Math.abs(ms);
  const time = {
    day: Math.floor(ms / 86400000),
    hour: Math.floor(ms / 3600000) % 24,
    minute: Math.floor(ms / 60000) % 60,
    second: Math.floor(ms / 1000) % 60,
    millisecond: Math.floor(ms) % 1000,
  };
  return time;
  // return Object.entries(time)
  //   .filter((val) => val[1] !== 0)
  //   .reduce((res, [key, value]) => ((res[key] = value), res), {} as any);
}
const padLeft = (str: string | number, num = 2, fill = '0') => String(str).padStart(num, fill);

export function formatDate(timeData: TimeDataType, formatStr: string, tag?: string) {
  const time = {
    DD: padLeft(timeData.day),
    D: padLeft(timeData.day, 1),
    HH: padLeft(timeData.hour),
    H: padLeft(timeData.hour, 1),
    mm: padLeft(timeData.minute),
    m: padLeft(timeData.minute, 1),
    ss: padLeft(timeData.second),
    s: padLeft(timeData.second, 1),
    SSS: padLeft(timeData.millisecond, 3),
  };
  if (tag === 'group') {
    const res: { [P in keyof TimeDataType]?: string } = {};
    const keyMap = new Map();
    keyMap.set('DD', 'day');
    keyMap.set('D', 'day');
    keyMap.set('HH', 'hour');
    keyMap.set('H', 'hour');
    keyMap.set('mm', 'minute');
    keyMap.set('m', 'minute');
    keyMap.set('ss', 'second');
    keyMap.set('s', 'second');
    keyMap.set('SSS', 'millisecond');
    // 此处需要保证循环顺序
    keyMap.forEach((value, key) => {
      if (formatStr.includes(key) && !res[value as keyof TimeDataType]) {
        res[value as keyof TimeDataType] = time[key as keyof typeof time];
      }
    });
    return res;
  }
  return formatStr.replace(
    new RegExp(`${Object.keys(time).join('|')}`, 'g'),
    subStr => time[subStr as keyof typeof time] || '',
  );
}

export function isDifferentTime(time1: number, time2: number, formatStr: string) {
  const actionsMap = new Map([
    ['SSS', () => time1 !== time2],
    ['s', () => Math.floor(time1 / 1000) !== Math.floor(time2 / 1000)],
    ['m', () => Math.floor(time1 / 60000) !== Math.floor(time2 / 60000)],
    ['H', () => Math.floor(time1 / 3600000) !== Math.floor(time2 / 3600000)],
    ['D', () => Math.floor(time1 / 86400000) !== Math.floor(time2 / 86400000)],
  ]);
  // eslint-disable-next-line no-restricted-syntax
  for (const [key, value] of actionsMap) {
    if (formatStr.includes(key)) {
      return value();
    }
  }
  return true;
}
