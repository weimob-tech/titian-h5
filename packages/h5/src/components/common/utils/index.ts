import { getTitianMode } from '../../../global/global';
import { IAnyObject, OmitType } from '../interface/index';

export * from './number';

export const objectToString = Object.prototype.toString;

export const toTypeString = (val: unknown): string => objectToString.call(val);

export type Key = string | number | symbol;

export const isString = (obj: unknown): obj is string => toTypeString(obj) === '[object String]';

export const isPlainObject = (obj: unknown): obj is { [key: Key]: unknown } => toTypeString(obj) === '[object Object]';

export const isPlainArray = (obj: unknown): obj is unknown[] => toTypeString(obj) === '[object Array]';

export const isPlainRegExp = (obj: unknown): obj is RegExp => toTypeString(obj) === '[object RegExp]';

export const isBoolean = (obj: unknown): obj is boolean => toTypeString(obj) === '[object Boolean]';

export const isFunction = (obj: unknown): obj is (...agrn: unknown[]) => unknown => typeof obj === 'function';

export const isArrayEqual = (arr1: unknown[], arr2: unknown[]): boolean => {
  if (arr1.length !== arr2.length) {
    return false;
  }

  return arr1.every((item, index) => item === arr2[index]);
};

export function debounce(fn: (...agrn: unknown[]) => unknown, ms = 300) {
  let timeoutId: ReturnType<typeof setTimeout>;
  return function debounceFn(this: unknown, ...args: unknown[]) {
    clearTimeout(timeoutId);
    timeoutId = setTimeout(() => fn.apply(this, args), ms);
  };
}

export function throttle(fn: (...agrn: unknown[]) => unknown, ms = 300) {
  let inThrottle: boolean;
  let lastFn: ReturnType<typeof setTimeout>;
  let lastTime: number;
  return function r(this: unknown, ...args: unknown[]) {
    if (!inThrottle) {
      fn.apply(this, args);
      lastTime = Date.now();
      inThrottle = true;
    } else {
      clearTimeout(lastFn);
      lastFn = setTimeout(() => {
        if (Date.now() - lastTime >= ms) {
          fn.apply(this, args);
          lastTime = Date.now();
        }
      }, Math.max(ms - (Date.now() - lastTime), 0));
    }
  };
}

export const randomString = () => Math.random().toString(36).slice(2);

export function filterInvalidData<T extends IAnyObject>(obj: T): OmitType<T, null | undefined> {
  const params = { ...obj };
  Object.keys(obj).forEach(key => {
    const value = obj[key];
    if (typeof value === 'undefined' || value === null) {
      delete params[key];
    }
  });
  return params;
}

export const DISABLED = 'disabled';

export function isDisabled(params: unknown): boolean {
  if (typeof params !== 'object' || !params) {
    return false;
  }
  return (params as IAnyObject)[DISABLED] as boolean;
}

export function range(num: number, min: number, max: number) {
  return Math.min(Math.max(num, min), max);
}

export function padZero(val: string | number, number = 2) {
  return `${val}`.padStart(number, '0');
}

export function isValidDate(d: unknown): d is Date {
  return d !== undefined && d !== null && !Number.isNaN(new Date(d as Date).getTime());
}

export function getDate(): Date;
export function getDate(date: string | number | Date): Date;
export function getDate(year: number, month: number): Date;
export function getDate(year: number, month: number, day: number): Date;
export function getDate(date?: string | number | Date, month?: number, day?: number) {
  let cur: Date;
  if (typeof date === 'undefined') {
    return new Date();
  }
  if (typeof date === 'number' && typeof month === 'number' && typeof day === 'number') {
    cur = new Date(date, month, day);
  } else if (typeof date === 'number' && typeof month === 'number') {
    cur = new Date(date, month);
  } else {
    cur = new Date(date);
  }
  if (!cur) {
    throw new Error('时间格式错误');
  }
  if (cur.toString() === 'Invalid Date') {
    throw new Error('时间格式错误');
  }
  return cur;
}

export function isUndefined(val: unknown) {
  return toTypeString(val) === '[object Undefined]';
}

// eslint-disable-next-line @titian-design/no-more-args
export function easeInOutCubic(t: number, b: number, c: number, d: number) {
  const cc = c - b;
  t /= d / 2;
  if (t < 1) {
    return (cc / 2) * t * t * t + b;
  }
  return (cc / 2) * ((t -= 2) * t * t + 2) + b;
}

// eslint-disable-next-line @titian-design/no-more-args
export function addEventListener<K extends Element>(
  target: K,
  eventType: string,
  cb: (arg: any) => any,
  options?: boolean | AddEventListenerOptions,
): () => void {
  if (target.addEventListener) {
    target.addEventListener(eventType, cb, options);
    document.addEventListener(eventType, cb, options);
  }

  return () => {
    if (target.removeEventListener) {
      target.removeEventListener(eventType, cb, options);
    }
  };
}

export const attrStyleToString = (attr: unknown) => {
  if (!attr) {
    return '';
  }
  if (typeof attr === 'string') {
    return attr;
  }
  if (Array.isArray(attr)) {
    return attr.join(';');
  }
  if (Object.keys(attr as { [key: string]: unknown }).length > 0) {
    return Object.keys(attr as { [key: string]: unknown })
      .reduce((array, key) => {
        array.push(`${key}:${(attr as { [key: string]: unknown })[key]}`);
        return array;
      }, [] as string[])
      .join(';');
  }

  return '';
};

export const stringToAttrStyle = (attrStr: unknown): { [key: string]: string } => {
  if (!attrStr) {
    return {};
  }
  if (isPlainObject(attrStr)) {
    return attrStr as { [key: string]: string };
  }
  if (!isString(attrStr)) {
    return {};
  }
  try {
    attrStr = JSON.parse(attrStr);
    // eslint-disable-next-line no-empty
  } catch (e) {}
  if (isPlainObject(attrStr)) {
    return attrStr as { [key: string]: string };
  }
  if (!isString(attrStr)) {
    return {};
  }
  return attrStr
    .trim()
    .split(';')
    .filter(Boolean)
    .reduce((pre, cur) => {
      const [key, value] = cur.split(':');
      const nkey = key.trim();
      pre[nkey] = value;
      return pre;
    }, {} as { [key: string]: string });
};

export const getResponsivePixel = (px: number) => {
  const mode = getTitianMode();
  if (mode === 'pc') {
    return px / 2;
  }
  return (document.documentElement.clientWidth / 750) * px;
};

export function getStyle<T extends HTMLElement>(element: T, attrName: keyof CSSStyleDeclaration) {
  let ret = element.style[attrName];
  if (ret) {
    return ret;
  }

  const getComputedStyle = document?.defaultView?.getComputedStyle;

  if (getComputedStyle) {
    ret = getComputedStyle(element, null)[attrName];
  }
  return ret;
}

export function setStyle(
  dom: HTMLElement,
  style: Partial<{ [key in keyof CSSStyleDeclaration]: CSSStyleDeclaration[key] | null }>,
) {
  (Object.keys(style) as (keyof CSSStyleDeclaration)[]).forEach(key => {
    if (style[key] === null) {
      dom.style.removeProperty(key as string);
    } else {
      dom.style.setProperty(key as string, style[key] as string);
    }
  });
}
export function hasClass(target: Element, className: string) {
  return target.className.indexOf(className) > -1;
}
export function addClass(target: Element, className: string) {
  return target.classList.add(className);
}
export function removeClass(target: Element, className: string) {
  return target.classList.remove(className);
}
export function toggleClass(target: Element, className: string) {
  return target.classList.toggle(className);
}
export function isCustomEvent(event: CustomEvent | Event): event is CustomEvent {
  return !event.isTrusted;
}

//  end 向后 无法查找 则  start end 之间
export function nextAvailable<T>(params: {
  start: number;
  end: number;
  list: T[];
  fn?: (list: T[], num: number) => boolean;
}) {
  const { start, end, list, fn } = params;
  let num = end;
  while (num < list.length) {
    num += 1;
    if (typeof list[num] !== undefined) {
      if (typeof fn === 'function' && fn(list, num)) {
        return num;
      }
    }
  }
  num = end;
  while (num > start) {
    num -= 1;
    if (typeof list[num] !== undefined) {
      if (typeof fn === 'function' && fn(list, num)) {
        return num;
      }
    }
  }
  return -1;
}

//  end 向前 无法查找 则 start end 之间
export function prevAvailable<T>(params: {
  start: number;
  end: number;
  list: T[];
  fn?: (list: T[], num: number) => boolean;
}) {
  const { start, list, end, fn } = params;
  let num = end;
  // 节点向前
  while (num > 0) {
    num -= 1;
    if (typeof list[num] !== undefined) {
      if (typeof fn === 'function' && fn(list, num)) {
        return num;
      }
    }
  }
  num = end;
  while (num < start) {
    num += 1;
    if (typeof list[num] !== undefined) {
      if (typeof fn === 'function' && fn(list, num)) {
        return num;
      }
    }
  }
  return -1;
}

export function addShadowRootStyle() {
  if (!this.extCss || !this.host || !this.host.shadowRoot) return;
  const style = this.host.shadowRoot.querySelector('style') || document.createElement('style');
  style.innerHTML = this.extCss;
  this.host.shadowRoot.appendChild(style);
}

export function pxToVW(str) {
  const mode = getTitianMode();

  if (mode === 'pc') return str;

  const res = (Number(str.replace('px', '')) * 100) / 750;
  function toFixedFn(value, n) {
    return Math.round(value * 10 ** n) / 10 ** n;
  }
  return `${toFixedFn(res, 5)}vw`;
}

export function validateEvent(event, shadowRoot) {
  const root = event.target.getRootNode();
  if (root === shadowRoot) {
    return true;
  }
  const slot = event.path.find(element => element.tagName === 'SLOT');
  const [node] = slot.assignedNodes();
  if (node && node.hasAttribute('slot')) {
    const stop = node.getAttribute('stopPropagation') || node.getAttribute('stop-propagation');
    if (typeof stop === 'boolean' && stop) {
      return false;
    }
    if (typeof stop === 'string' && (stop === 'true' || stop === '')) {
      return false;
    }
  }
  return true;
}

export function checkSlotIsUsed(host) {
  const slot = host.shadowRoot?.querySelector('slot')?.assignedNodes({ flatten: true });
  if (!slot || slot.length === 0) return false;
  if (slot && slot.length === 1) {
    if (slot[0].nodeName === '#text' && slot[0].textContent === '') {
      return false;
    }
  }
  return true;
}
