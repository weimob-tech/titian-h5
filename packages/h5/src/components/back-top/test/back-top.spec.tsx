import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { TiBackTop } from '..';
import { sleep } from '../../common/test';
import { TestContainer } from '../../common/test/container';
import * as scroll from '../../common/utils/scroll';

describe('渲染回到顶部', () => {
  beforeAll(() => {
    jest.spyOn(scroll, 'getScroll').mockReturnValue(1000);
  });

  afterAll(() => {
    jest.resetAllMocks();
  });
  it('简单使用', async () => {
    const page = await newSpecPage({
      components: [TiBackTop, TestContainer],
      template: () => (
        <test-container>
          <div style={{ height: '2000px' }}>
            <ti-back-top />
          </div>
        </test-container>
      ),
    });

    const backTopEl = page.root.querySelector('ti-back-top');
    const backTop = backTopEl.shadowRoot.querySelector<HTMLDivElement>('.titian-back-top');

    backTop.ownerDocument.dispatchEvent(new Event('scroll'));

    backTopEl.setAttribute('extCss', '.a {background: red}');
    await page.waitForChanges();
    await sleep();
    expect(page.root).toMatchSnapshot();

    backTop.click();
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('使用时展示回到顶部文字', async () => {
    const page = await newSpecPage({
      components: [TiBackTop, TestContainer],
      template: () => (
        <test-container>
          <div style={{ height: '2000px' }}>
            <ti-back-top text="顶部" />
          </div>
        </test-container>
      ),
    });

    page.win.scrollTo({ top: 1000 });
    await page.waitForChanges();
    await sleep(3000);

    expect(page.root).toMatchSnapshot();
  });

  it('删除节点，移除点击事件', async () => {
    const page = await newSpecPage({
      components: [TiBackTop, TestContainer],
      template: () => (
        <test-container>
          <ti-back-top text="顶部" />
        </test-container>
      ),
    });
    const backTopEl = page.root.querySelector('ti-back-top');

    page.root.removeChild(backTopEl);

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });
});
