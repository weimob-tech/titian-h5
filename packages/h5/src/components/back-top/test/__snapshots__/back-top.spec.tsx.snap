// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`渲染回到顶部 使用时展示回到顶部文字 1`] = `
<test-container>
  <!---->
  <div style="height: 2000px;">
    <ti-back-top>
      <mock:shadow-root>
        <div aria-hidden="true" class="titian-back-top" style="display: none;">
          <ti-icon name="go-to-top" size="32"></ti-icon>
          <span class="titian-back-top-text">
            顶部
          </span>
        </div>
      </mock:shadow-root>
    </ti-back-top>
  </div>
</test-container>
`;

exports[`渲染回到顶部 使用时展示回到顶部文字 2`] = `
<test-container>
  <!---->
  <div style="height: 2000px;">
    <ti-back-top extcss=".a {background: red}">
      <mock:shadow-root>
        <div aria-hidden="true" class="titian-back-top" style="display: none;">
          <ti-icon name="go-to-top" size="48"></ti-icon>
          <slot></slot>
        </div>
      </mock:shadow-root>
    </ti-back-top>
  </div>
</test-container>
`;

exports[`渲染回到顶部 使用时展示回到顶部文字 3`] = `
<test-container>
  <!---->
  <div style="height: 2000px;">
    <ti-back-top>
      <mock:shadow-root>
        <div aria-hidden="true" class="enterActiveClass enterDoneClass titian-back-top titian-slide-up-enter-active titian-slide-up-enter-done">
          <ti-icon name="go-to-top" size="32"></ti-icon>
          <span class="titian-back-top-text">
            顶部
          </span>
        </div>
      </mock:shadow-root>
    </ti-back-top>
  </div>
</test-container>
`;

exports[`渲染回到顶部 删除节点，移除点击事件 1`] = `
<test-container>
  <!---->
</test-container>
`;

exports[`渲染回到顶部 简单使用 1`] = `
<test-container>
  <!---->
  <div style="height: 2000px;">
    <ti-back-top extcss=".a {background: red}">
      <mock:shadow-root>
        <div aria-hidden="true" class="titian-back-top" style="display: none;">
          <ti-icon name="go-to-top" size="48"></ti-icon>
          <slot></slot>
        </div>
      </mock:shadow-root>
    </ti-back-top>
  </div>
</test-container>
`;

exports[`渲染回到顶部 简单使用 2`] = `
<test-container>
  <!---->
  <div style="height: 2000px;">
    <ti-back-top extcss=".a {background: red}">
      <mock:shadow-root>
        <div aria-hidden="true" class="enterActiveClass enterDoneClass titian-back-top titian-slide-up-enter-active titian-slide-up-enter-done">
          <ti-icon name="go-to-top" size="48"></ti-icon>
          <slot></slot>
        </div>
      </mock:shadow-root>
    </ti-back-top>
  </div>
</test-container>
`;
